### PR TITLE
Merge Search & State into new "Generator" type

### PR DESCRIPTION
**Pull request by RyanUnderhill**
```
  // Original version
  auto search = params.CreateSearch();
  auto state = model->CreateState(search->GetSequenceLengths(), params);

  while (!search->IsDone()) {
    search->SetLogits(state->Run(search->GetSequenceLength(), search->GetNextTokens()));
    search->SelectTop();
  }
```

```
  // Generator version
  auto generator = model->CreateGenerator();

  while (!generator->IsDone()) {
    generator->ComputeLogits();
    generator->AppendNextToken_Top();
  }
```

### CODE REVIEW

File: examples/llama.py [hunk 1]
```diff
@@ -10,48 +10,45 @@
...
 
 # Keep asking for input prompts in an loop
 while True:
     text = input("Input:")
-    # input_tokens = tokenizer.encode(text, return_tensors='np')
-    input_tokens = tokenizer.encode(text)
+    input_tokens = tokenizer.encode(text, return_tensors='np')
+    # input_tokens = tokenizer.encode(text)
 
     params=og.SearchParams(model)
```
Comment (yufenglee) [inline examples/llama.py:25]: could you please also change SearchParams to GeneratorParams?

File: examples/llama.py [hunk 2]
```diff
@@ -10,48 +10,45 @@
...
-    while not search.IsDone():
-        search.SetLogits(state.Run(search.GetSequenceLength(), search.GetNextTokens()))
+    while not generator.IsDone():
+        generator.ComputeLogits()
 
         # search.Apply_MinLength(1)
         # search.Apply_RepetitionPenalty(1.0)
 
-        search.SampleTopP(0.7, 0.6)
+        generator.AppendNextToken_TopP(0.7, 0.6)
```
Comment (yufenglee) [inline examples/llama.py:40]: Append doesn't look like a good name to me. generator actually generate next token.

Comment (yufenglee) [reply examples/llama.py:40]: And can we make TopP a parameter instead of part of the name?

File: examples/llama.py [hunk 3]
```diff
@@ -10,48 +10,45 @@
...
     print("Output:")
 
     print(text, end='', flush=True)
-    while not search.IsDone():
-        search.SetLogits(state.Run(search.GetSequenceLength(), search.GetNextTokens()))
+    while not generator.IsDone():
+        generator.ComputeLogits()
 
         # search.Apply_MinLength(1)
         # search.Apply_RepetitionPenalty(1.0)
```
Comment (yufenglee) [inline examples/llama.py:38]: could you please update this also

File: examples/llama.py (hunk 2 above)
Comment (RyanUnderhill) [reply examples/llama.py:40]: I named it that since you first compute the logits, then append the next token based on the logits.
I initially had AddNextToken but 'Add' was less correct than 'Append' since it's appending tokens, not adding them.
Did you have a better name?

File: examples/llama.py [hunk 4]
```diff
@@ -10,48 +10,45 @@
...
 
-        search.SampleTopP(0.7, 0.6)
+        generator.AppendNextToken_TopP(0.7, 0.6)
 
-        print(tokenizer.decode([search.GetNextTokens().GetArray()[0]]), ' ', end='', flush=True)
-        '''
+        # print(tokenizer.decode([generator.GetNextTokens().GetArray()[0]]), ' ', end='', flush=True)
         # Print each token as we compute it, we have to do some work to get newlines & spaces to appear properly:
-        word=tokenizer.convert_ids_to_tokens([search.GetNextTokens().GetArray()[0]])[0]
+        word=tokenizer.convert_ids_to_tokens([generator.GetNextTokens().GetArray()[0]])[0]
```
Comment (yufenglee) [inline examples/llama.py:44]: GetNexTokens looks fine. Is it better to use GetNewGeneratedTokens() considering that the tokens are already generated?

Comment (yufenglee) [reply examples/llama.py:44]: And we can make Generate(Append)NextTokens to return new generated tokens too

File: src/generators.h [hunk 5]
```diff
@@ -124,6 +104,52 @@ struct SearchParams {
   std::variant<Whisper> inputs;
 };
 
+struct Search {
```
Comment (yufenglee) [inline src/generators.h:107]: rename it to Sample?

File: examples/llama.py (hunk 2 above)
Comment (yufenglee) [reply examples/llama.py:40]: It is a process of generate next token. Add or Append doesn't reflect the action, i think.

File: src/generators.h (hunk 5 above)
Comment (RyanUnderhill) [reply src/generators.h:107]: This code does the Greedy/Beam search, the sampling is just a tiny part of it. But I might be able to hide it from generators.h

File: examples/llama.py (hunk 4 above)
Comment (RyanUnderhill) [reply examples/llama.py:44]: GetLastTokens()? Then it works in every case.

**Review by aciddelgado — APPROVED** [review]
